mlflow/models/evaluation/default_evaluator.py

```diff
@@ -0,0 +1,427 @@
+            raw_model = None
+            _logger.warning(f'Raw model resolution fails unexpectedly on PyFuncModel {model!r}, '
+                            f'error message is {e}')
+
+        if raw_model:
+            predict_fn = raw_model.predict
+            try:
+                import xgboost
+                if isinstance(raw_model, xgboost.XGBModel):
+                    # Because shap evaluation will pass evaluation data in ndarray format
+                    # (without feature names), if set validate_features=True it will raise error.
+                    predict_fn = partial(predict_fn, validate_features=False)
+            except ImportError:
+                pass
+
+        truncated_feature_names = [truncate_str_from_middle(f, 20) for f in feature_names]
+        for i, truncated_name in enumerate(truncated_feature_names):
+            if truncated_name != feature_names[i]:
+                # For truncated name, attach "(f_{feature_index})" at the end
+                truncated_feature_names[i] = f'{truncated_name}(f_{i})'
+
+        truncated_feature_name_map = {f: f2 for f, f2 in zip(feature_names, truncated_feature_names)}
+        if isinstance(X, pd.DataFrame):
+            X = X.rename(columns=truncated_feature_name_map, copy=False)
+
+        sampled_X = shap.sample(X, sample_rows)
+        if algorithm:
+            if algorithm == 'sampling':
+                explainer = shap.explainers.Sampling(
+                    predict_fn, X, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(X, sample_rows)
+            else:
+                explainer = shap.Explainer(
+                    predict_fn, sampled_X, feature_names=truncated_feature_names, algorithm=algorithm
+                )
+                shap_values = explainer(sampled_X)
+        else:
+            maskers = shap.maskers.Independent(sampled_X)
+            if raw_model and not is_multinomial_classifier and \
+                    shap.explainers.Linear.supports_model_with_masker(raw_model, maskers):
+                explainer = shap.explainers.Linear(
+                    raw_model, maskers, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(sampled_X)
+            elif raw_model and not is_multinomial_classifier and \
+                    shap.explainers.Tree.supports_model_with_masker(raw_model, maskers):
+                explainer = shap.explainers.Tree(
+                    raw_model, maskers, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(sampled_X)
+            elif raw_model and shap.explainers.Additive.supports_model_with_masker(
+                    raw_model, maskers
+            ):
+                explainer = shap.explainers.Additive(
+                    raw_model, maskers, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(sampled_X)
+            else:
```

Comment:
@WeichenXu123 Can we let shap pick the best explainer for the raw model, rather than trying to decide for ourselves, or does that cause problems?